Enable Pending Points (#319)
emdgroup/baybe: docs/userguide/async.md

> # Asynchronous Workflows

Reviewer: Note: Since there are several comments by @AdrianSosic already regarding the user guide, I only considered his comments for now and will give the user guide a fresh review once those comments have been addressed. Ping me here once this has happened.

> Asynchronous workflows describe situations where the loop between measurements and
> recommendations is more complex and needs to incorporate various other aspects. These

Reviewer 2: I think better to refer to "measurement" and "recommendation" in terms of the *processes*, not the resulting quantities
```suggestion
Asynchronous workflows describe situations where the loop between measurement and
recommendation is more complex and needs to incorporate various other aspects. These
```

could for instance be:
- **Distributed workflows**: When recommendations are distributed across several

>   operators, e.g. at different locations or in several reactors, some experiments might

Reviewer 2: perhaps a generalize using a different word than "reactors"? Pieces of equipment? Measurement devices? Lab equipment? Maybe you have a good idea...

Author: this is a concrete example started with `e.g.`, no need to provide a general statement after that

>   have been started, but are not ready when the next recommendation is requested.

Reviewer 2: I think the idea of working with batches can be helpful for understanding here
```suggestion
  have been started, but are not ready when the next batch of recommendations is requested.
```

  Without further consideration, the algorithm would be likely to recommend the pending

>   measurements again, as it is unaware they were already started.

Reviewer 2: perhaps add rationale why this is the case -> because, after all, they have been considered to be most promising

Author: imo not very needed + find it hard to include that int his sentence in addition to the other stuff please make suggestion

Reviewer: Agree with Martin here, is imo not needed.

Author: adrians response pending 4w, 1 more day otherwise I will yank this comment

Reviewer 2: OK, if you both think it is obvious to the reader (hard for me to tell, but I always assume the worst 😄 ) then let's leave it. My only way to include it would have been in the direction of `Without further consideration, the algorithm would be likely to recommend the pending measurements again (since they were and still are considered most promising), as it is unaware they were already started.`

By the way, terminology: I notice that we are still talking about `pending measurement` in the userguide, not `pending experiment` like we do it in the code. Shall we harmonize?

Author: included `(since they were and still are considered most promising)` which hopefully should satisfy everyone

> - **Partial targets**: Relevant when dealing with multiple targets which require very
>   different amounts of time to measure. When a new recommendation is requested, the
>   targets of previously recommended points might only partially available, but that
>   info should ideally already be considered.

Reviewer 2: Adjust style to match other bullet
```suggestion
- **Partial targets**: When dealing with multiple targets that require very different
  amounts of time to measure, the targets of previously recommended points might only be
  partially available when requesting the next batch of recommendations. Still, these
  partial measurements should ideally be considered when generating the recommendations.
```

## Pending Measurements

> With *pending measurements* we mean recommendations whose measurement process has

Reviewer 2: I think this paragraph needs to a slight overhaul. The problem I currently see: Above, you nicely describe the two *very different* situations that can arise (but also somehow imply that there might be others, since you say "for instance" – do we actually want to keep this or make it stricter and specifically state that these are the two classes of scenarios?)

But now it becomes a bit unclear what the "pending measurements" approach actually solves. Correct me if I'm wrong, but I think we only address the "distributed workflows" class, while the other remains a problem, right? If so, we should clearly state is by saying that BayBE offers a solution to the former via "pending_measurements", not to the latter? But somehow you seem to imply that it also solves the second, since you say `A measurement is pending whenever one or more of its targets are not measured yet`.... I'm a bit confused, tbh

Reviewer: Regarding the first point: There might be more scenarios, right? So I'd keep the "for instance".

Reviewer: I think we just need a very clear definition of what `pending` means. In my interpretation, this would be `A point is called *pending* if the process of measuring the point has started, but is not finished before a subsequent recommendation is requested`. This should fit both cases, and we can then write that we support adding pending points in Distributed Workflow setting resp. in a "Level 2" setting. Maybe these levels should also be explicitly mentioned here to make this more clear?

Author: Userguide was reworked please reevaluate

Reviewer 2: added my comments, so all good from my end. Thanks for reworking, btw!

been started, but not yet completed by time of triggering the next set of
recommendations. This is not to be confused with a situation where the targets of a
recommendation have been partially measured. A measurement is *pending* whenever one or
more of its targets are not measured yet. Only once all the targets are available,
the data point can be removed from `pending_measurements` and be provided via 
`measurements`.

To avoid repeated recommendations in the above scenario, BayBE provides the 

> `pending_measurements` keyword. It is available in `.recommend` from

Reviewer 2: Perhaps say more generally that it's available wherever recommendations can be requested and also mention the `RecommenderProtocol`?

Author: done

[`Campaign`](baybe.campaign.Campaign) or any

> [`Recommender`](baybe.recommenders.base.RecommenderProtocol).

Reviewer 2: You could actually make the links point directly to the methods

Author: done

```{admonition} Supported Acquisition Functions
:class: important

> `pending_measurementes` is only supported by Monte Carlo acquisition functions, i.e. the

Reviewer 2: ```suggestion
`pending_measurementes` is only supported by Monte Carlo (MC) acquisition functions, i.e. the
```

ones that start with a `q` in their name. Attempting to use a non-MC acquisition

> function with `pending_measurements` will result in a 

Reviewer 2: ```suggestion
function with `pending_measurements` will result in an 
```

[`IncompatibleAcquisitionFunctionError`](baybe.exceptions.IncompatibleAcquisitionFunctionError).
```

```{admonition} Supported Recommenders
:class: important

> For technical reasons, not every recommender might be able to fully or even partially

Reviewer 2: ```suggestion
For technical reasons, not every recommender is able to fully or even partially
```

> For technical reasons, not every recommender is able to fully or even partially

Reviewer: I think the wording of `fully or even partially` is confusing since I would not know what "partially" should even mean in this context. Maybe rephrase into something like "Note that different recommenders handle pending points in different ways"?

Author: was changed

utilize `pending_measurements`. For instance, 
[`BotorchRecommender`](baybe.recommenders.pure.bayesian.botorch.BotorchRecommender)
takes all pending measurements into account, even if they do not match exactly with any

> point in the searchspace.

Reviewer 2: ```suggestion
point in the search space.
```

[Clustering recommenders](baybe.recommenders.pure.nonpredictive.clustering.SKLearnClusteringRecommender)

> only incorporate pending points which are exact searchspace matches by excluding them

Reviewer 2: ```suggestion
only incorporate pending points that are exact search space matches by excluding them
```

from the candidates. The 
[`RandomRecommender`](baybe.recommenders.pure.nonpredictive.sampling.RandomRecommender)
does not take pending points into consideration at all and will raise a correspoding
[warning](baybe.exceptions.UnusedObjectWarning).
```

Akin to `measurements` or `recommendations`, `pending_measurements` is a dataframe in

> experimental representation. In the following example, we get a set of recommendations,

Reviewer 2: I think if you specifically mention "experimental representation", we should add a link to where this is explaine (i.e. search space?)

Author: still unable to dos o, please provide instructions

response pending 4w, 1 more day otherwise I will yank this comment

Reviewer 2: There is also a way to use implicit target that are autogenerated for headers, but I also couldn't get it running quickly. I think we should start creating a sphinx/myst cheatsheet somewhere, because every time I want to reference something, I basically start my search from scratch. And I'm like 99% sure that I already had this running previously in some other context!

Anyway, as a workaround, let's just use an explicit anchor. You can write
`is a dataframe in [experimental representation](#DATA_REPRESENTATION)` and then just put `(DATA_REPRESENTATION)=` above the representation section. Have tested it and it works

Reviewer 2: yes, tested this exact situation

Author: ok thanks for pointing it out, includede it (but currently cant verify lcoally due to docbuild hickups). willr esolve once confirmed

add results for half of them, and start the next recommendation, marking the
other half pending:
```python
# Get a set of 10 recommendation
rec = campaign.recommend(batch_size=10)

# Split recommendations into two parts
rec_finished = rec.iloc[:5]
rec_pending = rec.iloc[5:]

# Add target measurements to the finished part. Here we add a random number
rec_finished["Target_max"] = 1337
campaign.add_measurements(rec_finished)

# Get the next set of recommendations, incorporating the still unfinished ones
# These will not include the pending recommendations again

> rec_next = campaign.recommend(10, pending_measurements=rec_pending)

Reviewer: I would like the example to also contain the case of then adding the results for points that were pending before.

Author: I dont see why this would e. helpful as it would simply be like the very basic example adding data and then doing a second iteration

Reviewer: Resolving here since there is another thread dedicated to this discussion.
